File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_travelling_salesman.cpp
#include <cstdio>
#include <queue>
#include <vector>
#include <algorithm>
using namespace std;
int n, m, k, a, b, tim[30002], ans; bool vis[30002];
vector<int> adj[30002];
queue<int> Q;
int main(){
    scanf("%d %d", &n, &m);
    for(int i=0; i<m; i++){
        scanf("%d %d", &a, &b);
        adj[a].push_back(b);
        adj[b].push_back(a);
    }
    scanf("%d", &k);
    for(int i=0; i<k; i++){
        scanf("%d", &a);
        Q.push(a); tim[a]=0; vis[a]=true;
    }
    while(!Q.empty()){
        int t=Q.front(); Q.pop();
        for(int i=0; i<adj[t].size(); i++){
            if(!vis[adj[t][i]]){
                tim[adj[t][i]]=tim[t]+1; ans=max(ans, tim[adj[t][i]]);
                vis[adj[t][i]]=true; Q.push(adj[t][i]);
            }
        }
    }
    printf("%d\n", ans);
}

File: computer3/Fall Term/Solutions/Class09_Nov_06/lvl3_travelling_salesman.java
import java.util.*;
public class graph3p1 {
	public static void main(String [] args){
		Scanner in = new Scanner(System.in);
		int N = in.nextInt(), M = in.nextInt();
		ArrayList [] graph = new ArrayList [N+1];
		for(int i=0; i<=N; i++)
			graph[i] = new ArrayList<>();
		for(int i=0; i<M; i++){
			int u=in.nextInt(), v=in.nextInt();
			graph[u].add(v); graph[v].add(u);
		}
		Queue<Integer> Q = new LinkedList<>();
		int [] time = new int [N+1];
		boolean [] vis = new boolean [N+1];
		int C = in.nextInt(), ans = 0;
		for(int i=0; i<C; i++){
			int u = in.nextInt();
			Q.add(u); vis[u]=true; time[u]=0;
		}
		while(!Q.isEmpty()){
			int u = Q.poll();
			for(int v: (ArrayList<Integer>) graph[u]){
				if(!vis[v]){
					Q.add(v); vis[v]=true; 
					time[v]=time[u]+1;
					ans = Math.max(ans, time[v]);
				}
			}
		}
		System.out.println(ans);
	}
}